protocols/netlink/netlink_reader.go

```diff
@@ -0,0 +1,248 @@
+package protocolnetlink
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/bio-routing/bio-rd/config"
+	bnet "github.com/bio-routing/bio-rd/net"
+	"github.com/bio-routing/bio-rd/route"
+	"github.com/bio-routing/bio-rd/routingtable"
+	"github.com/bio-routing/bio-rd/routingtable/filter"
+	log "github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+)
+
+// Constants for IP family
+const (
+	IPFamily4 = 4 // IPv4
+	IPFamily6 = 6 // IPv6
+)
+
+// NetlinkReader read routes from the Linux Kernel and propagates it to the locRIB
+type NetlinkReader struct {
+	options *config.Netlink
+	routingtable.ClientManager
+	filter *filter.Filter
+
+	mu     sync.RWMutex
+	routes []netlink.Route
+}
+
+// NewNetlinkReader creates a new reader object and returns the pointer to it
+func NewNetlinkReader(options *config.Netlink) *NetlinkReader {
+	nr := &NetlinkReader{
+		options: options,
+		filter:  options.ImportFilter,
+	}
+
+	nr.ClientManager = routingtable.NewClientManager(nr)
+
+	return nr
+}
+
+// Read reads routes from the kernel
+func (nr *NetlinkReader) Read() {
+	log.WithField("rt_table", nr.options.RoutingTable).Info("Started netlink server")
+
+	// Start fetching the kernel routes after the hold time
+	time.Sleep(nr.options.HoldTime)
+
+	for {
+		// Family doesn't matter. I only filter by the rt_table here
+		routes, err := netlink.RouteListFiltered(IPFamily4, &netlink.Route{Table: nr.options.RoutingTable}, netlink.RT_FILTER_TABLE)
+		if err != nil {
+			log.WithError(err).Panic("Failed to read routes from kernel")
+		}
+
+		nr.propagateChanges(routes)
+
+		nr.mu.Lock()
+		nr.routes = routes
+
+		log.Debugf("NetlinkRouteDiff: %d", len(route.NetlinkRouteDiff(nr.routes, routes)))
+		nr.mu.Unlock()
+
+		time.Sleep(nr.options.UpdateInterval)
+	}
+}
+
+// create a path from a route
+func createPathFromRoute(r *netlink.Route) (*route.Path, error) {
+	nlPath, err := route.NewNlPathFromRoute(r, true)
+
+	if err != nil {
+		return nil, fmt.Errorf("Error while creating path object from route object: %v", err)
+	}
+
+	return &route.Path{
+		Type:        route.NetlinkPathType,
+		NetlinkPath: nlPath,
+	}, nil
+}
+
+// propagate changes to all subscribed clients
+func (nr *NetlinkReader) propagateChanges(routes []netlink.Route) {
+	nr.removePathsFromClients(routes)
+	nr.addPathsToClients(routes)
+}
+
+// Add given paths to clients
+func (nr *NetlinkReader) addPathsToClients(routes []netlink.Route) {
+	// only advertise changed routes
+	nr.mu.RLock()
+	advertise := route.NetlinkRouteDiff(routes, nr.routes)
+	nr.mu.RUnlock()
+
+	for _, r := range advertise {
+		// Is it a BIO-Written route? if so, skip it, dont advertise it
```

Comment:
Move the condition into a function and make the name of the function express the comment.

Reply:
sounds like a good idea! Thanks